File: SDCDownload.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jan 27 18:03:42 2021

@author: fwd
"""

import requests
import time
import re
import os

def downloadFile(url,path):
    start = time.time()
    size = 0
    #proxy = 'hk-aga-o-1.v2fast.win'
    #proxy_support = {'http':'http://'+proxy,'https':'https://'+proxy,}
    response = requests.get(url,stream = True) # proxies = proxy_support
    chunk_size = 1024#每次块大小为1024
    content_size = int(response.headers['content-length'])
    if os.path.exists(path) and os.path.getsize(path) == content_size:
        print('该文件已下载')
    else:
        print("文件大小："+str(round(float(content_size/chunk_size/1024),4))+"[MB]")
        with open(path,'wb') as FileTemp:
            for data in response.iter_content(chunk_size=chunk_size):#每次只获取一个chunk_size大小
                FileTemp.write(data)#每次只写入data大小
                size = len(data)+size
                print('\r'+"已经下载："+" 【"+str(round(size/chunk_size/1024,2))+"MB】"\
                      +"【"+str(round(float(size/content_size)*100,2))+"%"+"】",end="")
        end = time.time()
        print('\r'+"总耗时:"+str(end-start)+"s"+"平均速度："+str(content_size/chunk_size/1024/(end-start))+"M/s")



if __name__ == '__main__':
    Date = '2019-04-29/2019-04-30'.split('/')
    inst_ids = '&instrument_ids=fgm'  #仪器
    drm = '&data_rate_mode=srvy,brst'   #模式
    desc = '&descriptors='   #描述符
    sc_ids = '&sc_ids='    

    for ic in range(1,5):  #卫星编号
        sc_ids += 'mms'+str(ic)+','
    sc_ids = sc_ids[:-1]
        
    para = [inst_ids,drm,desc,sc_ids]
    url = 'https://lasp.colorado.edu/mms/sdc/public/files/api/v1/file_names/science?start_date='\
        + Date[0] + '&end_date=' + Date[1]
    
    for i in para:
        if i.split('=')[1] != '':
            url += i
            
    headers = {'Proxy-Connection': 'keep-alive'}
    SourceFile = requests.get(url,headers=headers).text
    expr = re.compile(r'mms[1234]_(.+?)\.cdf')
    FileNames = re.finditer(expr,str(SourceFile))
    #print(FileNames)
    
    FileUrl = 'https://lasp.colorado.edu/mms/sdc/public/files/api/v1/download/science?file='
    #迭代器在使用之前无法知道个数，不知道有没有更好的办法获取总数
    FileNum = len(re.findall(expr,SourceFile))    
    flag = 0
    path = "C:/Users/fwd/Desktop/python exercise/MMS/"
    for File in FileNames:
        flag += 1
        FileName = str(File.group())
        print("开始下载文件"+str(flag)+'/'+str(FileNum))
        downloadFile(FileUrl+FileName,path+FileName)

File: SDCDownload.m
clear;
clc

Date = '2019-01-01/2020-01-30';
splitDate = regexp(Date,'/','split');
% url = ['https://cdaweb.gsfc.nasa.gov/pub/data/mms/mms1/fgm/srvy/l2/',splitDate{1},'/',splitDate{2},'/'];
instrument_ids = 'fgm'; %仪器（可选）
data_rate_modes = 'srvy'; %模式（可选）
descriptors = ''; %描述符（可选,注意，如ins_ids中某仪器无该描述符，则不会下载该仪器的数据）
sc_ids = '';
for ic = 1:4    %卫星编号（可选）
    sc_ids = ['mms',num2str(ic),',',sc_ids];
end
if isempty(instrument_ids) == 0
    instrument_ids = ['&instrument_ids=',instrument_ids];
end
if isempty(data_rate_modes) == 0
    data_rate_modes = ['&data_rate_modes=',data_rate_modes];
end
if isempty(descriptors) == 0
    descriptors = ['&descriptors=',descriptors];
end
sc_ids=['&sc_ids=',sc_ids(1:end-1)];
url = ['https://lasp.colorado.edu/mms/sdc/public/files/api/v1/file_names/science?start_date=', ...
    splitDate{1},'&end_date=',splitDate{2},sc_ids,instrument_ids,data_rate_modes,descriptors];

sourcefile = webread(url);

% expression = '<a.+?href=\"(.+?)\">(.+?)</a>'; %识别网页中的链接
expression = 'mms[1234]_(.+?)\.cdf'; %识别cdf文件名
filenames = regexp(sourcefile,expression,'match');
filenames = unique(filenames);
% celldisp(filenames)

output_dir = ['D:\MMS\',splitDate{1},'To',splitDate{2},'\'];
if isfolder(output_dir) == 0
    mkdir(output_dir);
end
h = waitbar(0,'开始下载');
i = 1;
while i <= length(filenames)
    %检查文件是否已下载过（只能检查断点续传，即未改变下载参数情况下使用该程序所下过的，否则可能会缺漏）     
    if i < length(filenames)
        while isfile([output_dir,filenames{i+1}]) == 1 
            i = i + 1;
            s1 = ['文件夹中已有文件:',num2str(i),'/',num2str(length(filenames))];
            waitbar(i/length(filenames),h,s1);
        end
    end
    
    url_file = ['https://lasp.colorado.edu/mms/sdc/public/files/api/v1/download/science?', ...
        'file=',filenames{i}];
    output_filename = [output_dir,filenames{i}];   
    
    %网站接口需挂vpn下载，如果matlab下载速度过慢则需先挂vpn再打开matlab  
    options = weboptions('Timeout',10);    
    tic 
    websave(output_filename,url_file,options) ;
    TimeInterval = toc;     
    Dir = dir(output_dir); 
    FileIndex = find(strcmp({Dir.name},filenames{i}));
    SizeOfFile = Dir(FileIndex).bytes;
    AverageSpeed = SizeOfFile/(TimeInterval*1024*1024);
    
    %显示进度条
    s2 = ['已下载文件数: ',num2str(i),'/',num2str(length(filenames)), ...
        char(13,10)','最近文件下载速度:',num2str(AverageSpeed),'M/s'];
    waitbar(i/length(filenames),h,s2);   
    i = i+1;
end

if isempty(filenames)
    waitbar(0,h,'无可下载项');
else
    waitbar(1,h,'下载完毕ヽ(✿ﾟ▽ﾟ)ノ');
end
pause(3)
close(h)